### PR TITLE
Sets mappable list-type metadata (e.g. column_names) from a tool wrapper

### PR DESCRIPTION
**Pull request by yhoogstrate**
Hi all,

I have been working on a tool that produces a tabular file, with columns named after the input files. There was a discussion on how to do this with column names and metadata seemed to be the most intuitive way: https://github.com/galaxyproject/tools-iuc/issues/332 . The first problem I encountered was that setting metadata via an action does not allow to set list:

```
<data format="tabular" name="output" label="Three column output" >
    <actions>
        <action name="column_names" type="metadata" default="col1,col2,col3" />
    </actions>
</data>
```

This results in the following three columns: `c`, `o` and `l`, since the default value seems to be parsed as a string and strings are treated as iterators. I have solved this with the piece of code in lib/galaxy/tools/parameters/output.py.

A small note to that, I modified `<action type="metadata" />` instead of creating a new datatype (e.g. type="metadata_list").

When I tested it I was able to get the following column names: `col1`, `col2` and `col3`. Then I wanted to use the names of the input files rather than fixed column names:

```
<data format="tabular" name="output" label="Tool that aggregates data from $input_files" >
    <actions>
        <action name="column_names" type="metadata" default="${','.join([input.name for input in $input_files ])}" />
    </actions>
</data>
```

I hoped that the default value would be automatically parsed, like the labels of `<data />`-elements in the output, but I got the following two column names:
`${'` and `'.join([input.name for input in $input_files ])}`.

To solve this I made chances to lib/galaxy/tools/actions/__init__.py, by adding the function `set_metadata_defaults()` which does the same type of mapping as is done for the labels.

I am happy to hear how other people think about this

### CODE REVIEW

**Comment by bgruening** [comment]
@yhoogstrate I like this and have a few use-cases for it already as you might know ;)

Is the following working with your adjustments?

```
    <actions>
        <action name="column_names" type="metadata" default="$input_files" />
    </actions>
```

I think this is the most intuitive syntax.

**Comment by yhoogstrate** [comment]
Hi @bgruening,

It's the same mapping as used in the `<data label="" />` tags, so $input_files gets remapped to the actual file names. In planemo you get columns like:

`/tmp/tmpnvwcfa/files/000/dataset_2.dat` `/tmp/tmpnvwcfa/files/000/dataset_3.dat`

and in galaxy something like:

`<galaxy database directoy>/files/000/dataset_2.dat` `<galaxy database directoy>/files/000/dataset_3.dat`

($input_files is a `<param type="data" type="multiple" />` in this case)

Edit: if you want to find a cleaner solution, it might be handy to add a variable like `$input_files.names` to such input types

**Comment by bgruening** [comment]
@galaxybot test this
@yhoogstrate `$input_files.names` is what we should add in the future

**Comment by yhoogstrate** [comment]
I am close to solving the `$input_files.names` but I need help to finish it.

When I added:

```
--- a/lib/galaxy/tools/wrappers.py
+++ b/lib/galaxy/tools/wrappers.py
@@ -268,6 +268,12 @@ class DatasetListWrapper( list, ToolParameterValueWrapper, HasDatasets ):
         if not isinstance(datasets, list):
             datasets = [datasets]

+        self.names = [dataset.name.replace(",","") for dataset in datasets]
+        if(len(self.names) > 1):
+            self.names = ",".join(self.names)
+        else:
+            self.names = self.names[0]
+
         def to_wrapper( dataset ):
             if hasattr(dataset, "element_identifier"):
                 element = dataset
```

and I run a tool using `default="$input_files.names" />` for the first time, it sets the metadata properly. But when I re-run the tool with other input files, the metadata of the first time is being used.

**Comment by bgruening** [comment]
@blankenberg maybe you can help here?

File: lib/galaxy/tools/actions/__init__.py [hunk 1]
```diff
@@ -500,6 +502,21 @@ def get_output_name( self, output, dataset, tool, on_text, trans, incoming, hist
...
+        <data format="tabular" name="output" label="Tabular output, aggregates data from individual_inputs" >
+            <actions>
+                <action name="column_names" type="metadata" default="${','.join([input.name for input in $individual_inputs ])}" />
+            </actions>
+        </data>
+        """
+        if output.actions:
+            for action in output.actions.actions:
+                if action.tag == "metadata":
+                    action.default = fill_template( action.default, context=params )
```
Comment (jmchilton) [inline lib/galaxy/tools/actions/__init__.py]: This is subject to race conditions and seems a bit off - there is just one action.default per tool right not per invocation of tool right?

Comment (yhoogstrate) [inline lib/galaxy/tools/actions/__init__.py]: Thanks for the review @jmchilton 

> there is just one action.default per tool

This explains why the metadata is being set to the metadata of the first invocation. Do you think it is possible to do this for each invocation anyway?

Comment (jmchilton) [inline lib/galaxy/tools/actions/__init__.py]: My instinct is that you need to move all this out of `galaxy/tool/actions/__init__.py` (happens before the job) and into `finish()` of `JobWrapper` in `jobs/__init__.py` (happen after the job), but I'm not sure exactly.

**Comment by jmchilton** [comment]
Since it is not ready to be merged, I am going to mark this a WIP. I still hope to find some time to figure out the correct way to pass this information around.

**Comment by yhoogstrate** [comment]
Thanks @jmchilton, I've been looking in that part of the code and I was not able to quickly figure it out myself.

We could merge the changes in `lib/galaxy/tools/parameters/output.py`, since that's functional and already a usefull addition. This allows me to continue on some tools at least.

I could then create an issue for the let's call them non static metadata columns, as defined by the input.

**Comment by jmchilton** [comment]
@yhoogstrate They look fine at first glance, if they are useful please open a PR and I'll plus 1 it.

**Comment by yhoogstrate** [comment]
Hey @jmchilton,

I think I have found the solution. If you want to test it, you could use the following test tool:

https://gist.githubusercontent.com/yhoogstrate/d4f5bd49da96763ca8b4/raw/020536136b8d9f61943568b13726fcde9b358a5d/paste.xml

If you `planemo serve` it and use the following input files:

https://gist.githubusercontent.com/yhoogstrate/a13fb71a0d3adc6b9351/raw/c468de76c509e8be111158a02cd0562f1c7b8fc3/t1.txt
https://gist.githubusercontent.com/yhoogstrate/8787dd7e072f0c734e6a/raw/0c9f07f1a065ce819b8e7282408e78d60abe7676/t2.txt
https://gist.githubusercontent.com/yhoogstrate/cb1527c0d973aa84b238/raw/7a0ad83f6dcde4039af5ac95e33bd567389267cd/t3.txt

You should see that the output files have new column names defined by metadata.

**Comment by jmchilton** [comment]
@galaxybot test this

**Comment by jmchilton** [comment]
This looks a lot a better, I think this is in pretty good shape now. I want to see the API tests pass (they will) and then I'll merge.

**Comment by jmchilton** [comment]
Thanks for sticking with this @yhoogstrate!

**Comment by yhoogstrate** [comment]
Thank you for the quick review, again!

**Pull request merged**